src/packages/storage-provider/src/decorators/media-field.ts

```diff
@@ -0,0 +1,134 @@
+import {
+	BaseDataEntity,
+	Entity,
+	Field,
+	FieldOptions,
+	GraphQLEntity,
+	graphweaverMetadata,
+	FieldMetadata,
+} from '@exogee/graphweaver';
+import { S3StorageProvider } from '../storageProvider';
+import { Source } from 'graphql';
+
+export interface MediaData extends BaseDataEntity {
+	filename: string;
+	type: MediaType;
+	url?: string;
+}
+
+export enum MediaType {
+	IMAGE = 'IMAGE',
+	OTHER = 'OTHER',
+}
+
+graphweaverMetadata.collectEnumInformation({
+	name: 'MediaType',
+	target: MediaType,
+});
+
+type MediaTypeFieldOptions = FieldOptions & {
+	storageProvider: S3StorageProvider;
+};
+
+const isMedia = (value: unknown): value is MediaData =>
+	!!(
+		value &&
+		value !== null &&
+		typeof value == 'object' &&
+		'filename' in value &&
+		typeof value.filename === 'string' &&
+		'type' in value &&
+		typeof value.type === 'string'
+	);
+
+@Entity('Media')
+export class Media extends GraphQLEntity<MediaData> {
+	public dataEntity!: MediaData;
+
+	@Field(() => String)
+	filename!: string;
+
+	@Field(() => MediaType)
+	type!: MediaType;
+
+	@Field(() => String, { apiOptions: { excludeFromBuiltInWriteOperations: true } })
+	url!: string;
+
+	static serialize = ({ value }: { value: unknown }) => {
+		if (value === null) return null;
+		if (isMedia(value)) {
+			return JSON.stringify(
+				{
+					filename: value.filename,
+					type: value.type,
+				},
+				null,
+				2
```

Comment:
I reckon we should save wire space here and remove the `null, 2` parameters. I know the DB will then store it however it wants, so no need to make it pretty on the way through I reckon.